include/up-cpp/transport/datamodel/UPayload.h

```diff
@@ -39,124 +39,118 @@ namespace uprotocol::utransport {
         UNDEFINED  /* invalid */
     };
 
+    // The Serialization format for the data stored in UPayload.
+    enum UPayloadFormat {
+        UNSPECIFIED = 0,             /* Payload format was not is not set */
+        PROTOBUF_WRAPPED_IN_ANY = 1, /* Payload is an Any protobuf message that contains the packed payload */
+        PROTOBUF = 2,                /* Protobuf serialization format */
+        JSON = 3,                    /* JSON serialization format */
+        SOMEIP = 4,                  /* Basic SOME/IP serialization format */
+        SOMEIP_TLV = 5,              /* SOME/IP TLV format */
+        RAW = 6,                     /* RAW (binary) format */
+        TEXT = 7                     /* Text format */
+    };          
+
     /**
     * The UPayload contains the clean Payload information at its raw serialized structure of a byte[]
     */
     class UPayload {
-        public:
 
-            UPayload(const uint8_t *data, 
-                     const size_t &dataSize,
-                     const UPayloadType &type) {
-                
-                payloadType_ = type;
+        public:
+            // Constructor
+            UPayload(const uint8_t* ptr, 
+                     const size_t size, 
+                     const UPayloadType &type) : dataSize_(size), type_(type) {
 
-                if ((nullptr == data) || (0 == dataSize)) {
-                    data_ = nullptr;
-                    refPtr_ = nullptr;
-                    dataSize_ = 0;
+                if (type == UPayloadType::REFERENCE) {
+                    dataPtr_ = std::shared_ptr<const uint8_t[]>(ptr, [](const uint8_t*){});
+                } else {
+                    dataPtr_ = std::shared_ptr<const uint8_t[]>(new uint8_t[dataSize_], [](const uint8_t* p){ delete[] p; });
+                    if (nullptr != ptr) {
+                        std::memcpy(const_cast<uint8_t*>(dataPtr_.get()), ptr, dataSize_);
+                    }
```

Comment:
Is this an interface that would be used by consumers of this library? I'm a little concerned about this interface/implementation (bare pointers with no ownership semantics, not using containers with included size tracking, allocating memory for a shared_ptr using new instead of make_shared, etc). If it's purely internal to the library and not meant to be used by consumers, then it's probably fine for now. Otherwise if developers will be writing code to this, we should probably work on a cleaner interface.

Reply:
UPayload used as a container , basically you set the buffer you want to send to UPayload.

Could you please provide a code sample for the changes ?

Reply:
[Here's an example](https://github.com/MelamudMichael/up-cpp/compare/feature/rpc_alignment...gregmedd:up-cpp:example/UPayload-without-bare-pointers)